compiler/rustc_middle/src/ty/relate.rs

```diff
@@ -740,6 +771,20 @@ impl<'tcx, T: Relate<'tcx>> Relate<'tcx> for ty::Binder<'tcx, T> {
     }
 }
 
+impl<'tcx, T: Relate<'tcx>> Relate<'tcx> for Option<T> {
+    fn relate<R: TypeRelation<'tcx>>(
+        relation: &mut R,
+        a: Self,
+        b: Self,
+    ) -> RelateResult<'tcx, Self> {
+        match (a, b) {
+            (None, None) => Ok(None),
+            (Some(a), Some(b)) => relation.relate(a, b).map(Some),
+            _ => Err(TypeError::Mismatch),
```

Comment:
Can you add a FIXME to remove this impl? 

It'll result in poor type errors since we're returning `Mismatch` instead of `Sorts`, and the match in `impl<'tcx> Relate<'tcx> for Pattern<'tcx> {` should just handle all the cases.

I don't think other code should be using this.

Reply:
Ping

Reply:
I just removed the impl and inlined it (still added a FIXME there)